plugin/pulpcore/plugin/stages/content_unit_stages.py

```diff
@@ -0,0 +1,189 @@
+    the db for efficiency.
+    """
+
+    async def __call__(self, in_q, out_q):
+        """
+        The coroutine for this stage.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+
+        Returns:
+            The coroutine for this stage.
+        """
+        batch = []
+        shutdown = False
+        while True:
+            try:
+                declarative_content = in_q.get_nowait()
+            except asyncio.QueueEmpty:
+                if not batch and not shutdown:
+                    declarative_content = await in_q.get()
+                    batch.append(declarative_content)
+                    continue
+            else:
+                batch.append(declarative_content)
+                continue
+
+            content_artifact_bulk = []
+            remote_artifact_bulk = []
+            remote_artifact_map = {}
+
+            with transaction.atomic():
+                for declarative_content in batch:
+                    if declarative_content is None:
+                        shutdown = True
+                        continue
+                    if declarative_content.content.pk is None:
+                            declarative_content.content.save()
+                            for declarative_artifact in declarative_content.d_artifacts:
+                                content_artifact = ContentArtifact(
+                                    content=declarative_content.content,
+                                    artifact=declarative_artifact.artifact,
+                                    relative_path=declarative_artifact.relative_path
+                                )
+                                content_artifact_bulk.append(content_artifact)
+                                remote_artifact_data = {
+                                    'url': declarative_artifact.url,
+                                    'size': declarative_artifact.artifact.size,
+                                    'md5': declarative_artifact.artifact.md5,
+                                    'sha1': declarative_artifact.artifact.sha1,
+                                    'sha224': declarative_artifact.artifact.sha224,
+                                    'sha256': declarative_artifact.artifact.sha256,
+                                    'sha384': declarative_artifact.artifact.sha384,
+                                    'sha512': declarative_artifact.artifact.sha512,
+                                    'remote': declarative_artifact.remote,
+                                }
+                                content_pk = content_artifact.content.pk
+                                remote_artifact_map[content_pk] = remote_artifact_data
```

Comment:
After switching to *integer* PK, the `content_artifact.content.pk` will be (None) because the model has not yet been saved.  Right?

Reply:
This is rebased onto the integer PK already so it uses them currently. The pk is set from L#151.

I verified I cannot use the models themselves because they are unsaved and Django raises an unhashable error. We could use sha512 instead. The only reason I didn't is to save memory. Should I switch it to that?